Assets/MixedRealityToolkit.SDK/Experimental/Features/Utilities/Migration/Tools/MigrationWindow.cs

```diff
@@ -0,0 +1,417 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using UnityEditor;
+using UnityEngine;
+using Microsoft.MixedReality.Toolkit.Utilities.Editor;
+using UnityEngine.SceneManagement;
+using UnityEditor.SceneManagement;
+
+using Object = UnityEngine.Object;
+
+namespace Microsoft.MixedReality.Toolkit.Experimental.Utilities
+{
+    /// <summary>
+    /// This tool allows the migration of game object obsolete components into up-to-date versions.
+    /// Requires Implementation of IMigrationhandler.
+    /// </summary>
+    public class MigrationWindow : EditorWindow
+    {
+        private enum ToolbarOption 
+        { 
+            Objects = 0, 
+            Scenes = 1, 
+            Project = 2 
+        };
+
+        private static readonly string[] toolbarTitles =
+{
+            "Objects",
+            "Scenes",
+            "Full Project"
+        };
+
+        private ToolbarOption selectedToolbar = ToolbarOption.Objects;
+        private Vector2 scrollPosition = Vector2.zero;
+
+        private const string DependencyWindowURL = "https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Tools/MigrationWindow.html";
+        private const string WindowTitle = "Migration Window";
+        private const string WindowDescription = "This tool allows the migration of obsolete components into up-to-date versions. Component-specific implementation of the MigrationHandler Interface is required.";
+
+        private const string SceneExtension = ".unity";
+        private const string PrefabExtension = ".prefab";
+
+        private Dictionary<string, Type> migrationTypesMap = new Dictionary<string, Type>();
+        private String[] migrationTypeNames;
+        private int selectedMigrationHandlerIndex;
+        private List<Object> migrationObjects = new List<Object>();
+        private Object selection;
+
+        private dynamic migrationTypeInstance
+        {
+            get
+            {
+                var selectedMigrationTypeName = migrationTypeNames[selectedMigrationHandlerIndex];
+                var migrationHandlerType = migrationTypesMap[selectedMigrationTypeName];
+                return Activator.CreateInstance(migrationHandlerType) as IMigrationHandler;
```

Comment:
@keveleigh this will create a instance of migrationTypeInstance each time this getter is called right? My C# isn't as strong as yours, but unless there's some magic here (where the runtime knows about/watches selectedMigrationHandlerIndex to know that things should only change based on that), I assume this would create a new instance of the handler on each.

If this is required, then we should actually document this (i.e. comment), because someone could easily look at this (i.e. what I'm doing) and be like "oh we don't need to do this, let's just cache it"

Reply:
will cache the instance